src/doorman/TelnetProtocol.hpp

```diff
@@ -0,0 +1,46 @@
+#pragma once
+
+#include "Byte.hpp"
+
+#include <gsl/span>
+
+#include <vector>
+
+class TelnetProtocol {
+public:
+    struct Handler {
+        virtual ~Handler() = default;
+        virtual void send_bytes(gsl::span<const byte> data) = 0;
+        virtual void on_line(std::string_view line) = 0;
+        virtual void on_terminal_size(int width, int height) = 0;
+        virtual void on_terminal_type(std::string_view type, bool ansi_supported) = 0;
+    };
+
+private:
+    Handler &handler_;
+    std::vector<byte> buffer_;
+    bool got_term_ = false;
+    bool echoing_ = true;
+    int width_ = 80;
+    int height_ = 24;
+    bool ansi_ = false;
+    char previous_separator_ = 0;
+
+    void send_com(byte a, byte b);
+    void send_opt(byte a);
+
+    size_t on_subcommand(gsl::span<const byte> command_sequence);
+    size_t on_command(gsl::span<const byte> command_sequence);
+    gsl::span<const byte> interpret_iacs();
+    std::string_view find_whole_lines(std::string_view data);
```

Comment:
I believe this function's name should be "handle_whole_lines". Find suggests it returns the whole lines (although the return type isn't consistent with that).

Maybe renaming to 'handle_whole_lines' and adding a comment explaining that unhandled trailing text from <data> is returned? Assuming I've understood the function's function correctly.

Reply:
Yeah this one's tricksie. I went back and forth a lot on this. Will rename and comment the return value.